File: Sources/XMTPiOS/CodecRegistry.swift
//
//  CodecRegistry.swift
//
//
//  Created by Pat Nakajima on 12/22/22.
//

import Foundation

class CodecRegistry {
	private let lock = NSLock()
	private var codecs: [String: any ContentCodec] = [
		TextCodec().id: TextCodec(),
	]

	func register(codec: any ContentCodec) {
		lock.lock()
		defer { lock.unlock() }
		codecs[codec.id] = codec
	}

	func find(for contentType: ContentTypeID?) -> any ContentCodec {
		lock.lock()
		defer { lock.unlock() }

		guard let contentType else {
			return TextCodec()
		}

		if let codec = codecs[contentType.id] {
			return codec
		}

		return TextCodec()
	}

	func find(for contentTypeString: String) -> any ContentCodec {
		lock.lock()
		defer { lock.unlock() }

		for (_, codec) in codecs {
			if codec.description == contentTypeString {
				return codec
			}
		}

		return TextCodec()
	}
}

extension CodecRegistry {
	func isRegistered(codec: any ContentCodec) -> Bool {
		lock.lock()
		defer { lock.unlock() }
		return codecs[codec.id] != nil
	}

	func isRegistered(codecId: String) -> Bool {
		lock.lock()
		defer { lock.unlock() }
		return codecs[codecId] != nil
	}

	func removeCodec(for id: String) {
		lock.lock()
		defer { lock.unlock() }
		codecs.removeValue(forKey: id)
	}
}

File: sdks/android/shared/src/main/java/org/xmtp/kotlin/CodecRegistry.kt
package org.xmtp.kotlin

import org.xmtp.kotlin.codecs.ContentCodec
import org.xmtp.kotlin.codecs.ContentTypeId
import org.xmtp.kotlin.codecs.TextCodec
import org.xmtp.kotlin.codecs.id

data class CodecRegistry(
    val codecs: MutableMap<String, ContentCodec<*>> = mutableMapOf(),
) {
    fun register(codec: ContentCodec<*>) {
        codecs[codec.contentType.id] = codec
    }

    fun find(contentType: ContentTypeId?): ContentCodec<*> {
        contentType?.let {
            val codec = codecs[it.id]
            if (codec != null) {
                return codec
            }
        }
        return TextCodec()
    }

    fun findFromId(contentTypeString: String): ContentCodec<*> {
        for ((_, codec) in codecs) {
            if (codec.contentType.id == contentTypeString) {
                return codec
            }
        }
        return TextCodec()
    }
}
